src/System.Text.CaseFolding/System/Text/CaseFolding/SimpleCaseFolding.cs

```diff
@@ -0,0 +1,630 @@
+                                dst = ch1.lowSurrogate;
+                            }
+                            else
+                            {
+                                dst = c1;
+                                dst = ref Unsafe.Add(ref dst, 1);
+                                dst = c1Low;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        // No low surrogate - throw?
+                        // We would have to throw but we can do simple binary copy
+                        // because the simple case folding is only used for comparisons.
+                        dst = c1;
+
+                        // Undo because it can be regular char and we will process it in next cycle.
+                        src = ref Unsafe.Add(ref src, -1);
+                    }
+                }
+                else
+                {
+                    // We expect a high surrogate but get a low surrogate - throw?
+                    // We would have to throw but we can do simple binary copy
+                    // because the simple case folding is only used for comparisons.
+                    dst = c1;
+                }
+
+                // Move to next char.
+                length--;
+                src = ref Unsafe.Add(ref src, 1);
+                dst = ref Unsafe.Add(ref dst, 1);
+
+                // For char below 0x5ff use fastest 1-level mapping.
+                while (length != 0 && src <= MaxChar)
+                {
+                    dst = Unsafe.Add(ref refMapBelow5FF, src);
+                    src = ref Unsafe.Add(ref src, 1);
+                    dst = ref Unsafe.Add(ref dst, 1);
+                    length--;
+                }
+
+                if (length == 0)
+                {
+                    return;
+                }
+
+                // We catch a char above 0x5ff.
+                // Process it with more slow two-level mapping.
+                while (length != 0 && !IsSurrogate(src))
+                {
+                    var v1 =  Unsafe.Add(ref refMapLevel1, src >> 8);
+                    var ch1 = Unsafe.Add(ref refMapData, v1 + (src & 0xFF));
+                    if (ch1 == 0)
+                    {
+                        ch1 = src;
+                    }
+
+                    dst = ch1;
+                    src = ref Unsafe.Add(ref src, 1);
+                    dst = ref Unsafe.Add(ref dst, 1);
+                    length--;
+                }
+
+                if (length == 0)
+                {
+                    return;
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsAscii(char c)
+        {
+            return c < 0x80;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsHighSurrogate(char c)
```

Comment:
The code is very sensitive so I have to locally put the inlining methods. During the transfer to CoreFX, we could repeat perf measurements.

Reply:
To me, this just looks like unnecessary tech debt being built up without much benefit. I don't think we would ever have duplicate `Char.IsHighSurrogate` implementations lying around when we move to corefx just to avoid using an existing API. I am surprised to hear that the existing API is slow. It might be an artifact of what you are benchmarking. The implementations are essentially identical.

If you have strong evidence of a performance issue, we should then consider marking the existing `Char.IsHighSurrogate` as `AggressiveInlining`

cc @stephentoub

Reply:
If we will move this to CoreFX will there be standalone package? If yes maybe we don’t want any "external" dependencies. Also there already seems is some duplications in CoreFX (I found its by Unicode constant names).

Reply:
> If we will move this to CoreFX will there be standalone package?

I am not sure and we can defer that decision to API reviewers when the API is ready. Let's start with assuming it will ship as part of an existing assembly and not as a standalone package.

Also, I would not consider depending on `Char.IsHighSurrogate` as an external dependency (similar to types like string/char/etc.), regardless. They exist in corelib and this API in particular has existed since .net standard 1.0.